Feature/remove transactions queue 1027

sebastiandelgado opened the pull request:
Remove `PendingSignRequests` queue from the sign module.

This closes #1027 by removing the pending sign requests queue  dependency from the `SendTransaction`, `SignMessage` and `Recover`. It was split into 3 different PRs that were reviewed before merging into this feature branch:
 - Remove `pendingSignRequests` queue from the transactor and sign transactions without the queue on `SendTransaction` calls
 - Remove `pendingSignRequests` queue from the `personalAPI` service module so that messages are signed/recovered without it in the `SignMessage` and `Recover` methods
 - Remove all of the `pendingSignRequests` code, along with its references in `api/backend.go` and any tests that used it.

The `sign` module was left pretty barebones, so we might want to simplify that in a separate PR (this one is definitely big enough already) by removing most interfaces from them.

@mandrigin @adambabik Thanks for all the help reviewing these PRs. Let me know if you spot anything here before merging.

@flexsurfer please take a look at the `library.go` file which has all the changes to the rpc interface (basically add `SendTransaction`, `SignMessage` and `Recover` methods in place of the `ApproveSignRequest...` methods).

What considerations do I need to take when merging this? For example, if I need to hold off on merging this so you can update the status-react interface to match these changes. What do we need to coordinate with status-react so that we don't break anything when deploying this change?

Commit:
Remove PendingSignRequests queue from personalSign (#1102)

Commit:
Remove PendingSignRequests queue from Transactor (#1101)

mandrigin reviewed (approved):
LGTM

divan commented on api/backend.go:
```diff
@@ -299,26 +273,18 @@ func (b *StatusBackend) registerHandlers() error {
-
-	rpcClient.RegisterHandler(params.PersonalSignMethodName, b.personalAPI.Sign)
-	rpcClient.RegisterHandler(params.PersonalRecoverMethodName, b.personalAPI.Recover)
+	rpcClient.RegisterHandler(params.PersonalSignMethodName, unsupportedMethodHandler)
```
Why is this method not supported? Is it unsupported both with using infura and les?

adambabik replied:
Because we want to guard calling it with a password. It's not supported to be called directly, only via a binding. Maybe we can add a comment explaining that.

divan commented on services/personal/api.go:
```diff
@@ -22,37 +23,44 @@ var (
-	return &metadata{data, address}, nil
+// UnmarshalSignRPCParams puts the RPC params for `personal_sign` or `web3.personal.sign`
+// into SignParams
+func UnmarshalSignRPCParams(rpcParamsJSON string) (SignParams, error) {
```
It looks like the better approach would be implementing [json.Unmarshaler](https://golang.org/pkg/encoding/json/#Unmarshaler) interface for these types (`SignParams` and `RecoverParams`). So the simple `json.Unmarshal()` will do the job.

divan commented on services/personal/api.go:
```diff
@@ -22,37 +23,44 @@ var (
+}
+
+// UnmarshalRecoverRPCParams
+func UnmarshalRecoverRPCParams(rpcParamsJSON string) (RecoverParams, error) {
```
Same as previous comment.

divan commented on transactions/types.go:
```diff
@@ -89,3 +89,10 @@ func RPCCalltoSendTxArgs(args ...interface{}) (SendTxArgs, error) {
 }
+
+// UnmarshalSendTxRPCParams creates SendTxArgs based on RPC JSON
+func UnmarshalSendTxRPCParams(rpcParamsJSON string) (SendTxArgs, error) {
```
This is just a wrapper that runs `json.Unmarshal` and is used only once in `lib/library.go`. I would keep it on the caller side (i.e. in `lib/` instead).

adambabik commented on lib/library.go:
```diff
@@ -210,139 +212,59 @@ func Logout() *C.char {
-	outBytes, err := json.Marshal(out)
+// SignMessage unmarshals rpc params {data, address, password} and passes
+// them onto backend.SignMessage
+// nolint: deadcode
```
This line should be replaced with `//export SignMessage`. Otherwise, this function won't be public.

adambabik commented on lib/library.go:
```diff
@@ -210,139 +212,59 @@ func Logout() *C.char {
-	parsedIDs, err := ParseJSONArray(C.GoString(ids))
+// Recover unmarshals rpc params {signDataString, signedData} and passes
+// them onto backend.
+// nolint: deadcode
```
The same here `//export Recover`.

adambabik commented on lib/library.go:
```diff
@@ -210,139 +212,59 @@ func Logout() *C.char {
 
-	outBytes, err := json.Marshal(out)
+// SendTransaction converts RPC args and calls backend.SendTransaction
+// nolint: deadcode
```
And here `//export SendTransaction`.

adambabik commented on lib/types.go:
```diff
@@ -77,24 +77,9 @@ type NotifyResult struct {
 }
 
 // SignRequestResult is a JSON returned from transaction complete function (used in exposed method)
+// nolint: megacheck
```
I run `make lint` locally without this line and there were no errors.

sebastiandelgado commented:
@adambabik @divan Addressed all comments. Only one I addressed differently to the suggestions was the unmarshalling, where I got rid of the unmarshalling methods altogether and simply did the unmarshalling in the lib methods (`SendTransaction`, `Recover`, `SignMessage`)

adambabik reviewed (approved):
Great job! 👏

mandrigin commented:
@sebastiandelgado added a patch to Makefile to avoid some dep-related leftovers on Jenkins.

Commit:
Remove sign.PendingRequests queue (#1117)

sebastiandelgado commented:
This PR is rebased and ready to merge. Just waiting on confirmation that the interface works as expected before merging into develop

adambabik commented:
@sebastiandelgado can you please link a status-react PR with this change?

flexsurfer commented:
@adambabik there is no PR yet, i'm in progress, just started work

flexsurfer commented:
hey @sebastiandelgado `SignMessage` requires password in rpcParams, and ignores it in function params, also i need params to return to js callback with error and result, also because we don't have failed signal anymore we need this information here about error not only message but id

status-im-jenkins commented:
Can one of the admins verify this patch?

adambabik committed:
Merge branch 'develop' into feature/remove-transactions-queue-1027

adambabik committed:
remove sign package

adambabik committed:
fix lib/ unit tests

adambabik commented:
What left here is to return a specific error code for `keystore.ErrDecrypt`.

adambabik committed:
add error code for invalid password

adambabik commented:
@flexsurfer Codes for `eth_sendTransaction`, `personal_sign` and `personal_ecRecover`:
```
// special codes
codeFailedParseResponse = 1
codeFailedParseParams = 2
// account related codes
codeErrNoAccountSelected = 3
codeErrInvalidTxSender = 4
codeErrDecrypt = 5
```

adambabik committed:
Revert "Add build tags to signal package (#1119)"

flexsurfer commented:
https://docs.google.com/drawings/d/1l472xaRgEJzizxUkNmeL0STSNqX9ClUy0OoZpKEjjXo

new TX flow

adambabik committed:
Merge branch 'develop' into feature/remove-transactions-queue-1027

flexsurfer commented:
@adambabik before it was not possible to send the same tx twice because we had a queue on go side, do you have some protection in this PR from sending the same tx twice? for example, if we call `SendTransaction` twice you can't distinguish that this is the same tx and will send it twice?

adambabik commented:
@flexsurfer how is it possible to send two transactions? I though that after this change each `eth_sendTransaction` call will result in a blocking confirmation screen. Until it's canceled or confirmed it won't be possible to send another transaction. Is that not true?

From status-go point of view, each call of `SendTransaction` binding is a separate transaction, there is nothing to protect from sending the same tx twice (they will get different nonces) and there wasn't any protection before. With the queue mechanism, two transactions would be scheduled.

flexsurfer commented:
@adambabik we had a bug in the past, when it was possible to press sign button twice and we called approveTransaction twice, but we had tx id (because you had a queue on go side), nothing wrong happened , but now if we'll have same bug in status-react two transactions will be sent, so we need to be super careful on status-react side now

adambabik committed:
Merge branch 'develop' into feature/remove-transactions-queue-1027

flexsurfer reviewed (approved):
tested, ready to be merged!

adambabik commented:
@flexsurfer success! Check out https://ci.status.im/job/status-go/job/status-go-manual/302/consoleFull. I built it from a tag, can you check if `tags-v0.11.0^0-g4afd9e6c-302` is gonna work? It has a strange artefact :/

<blockquote><img src="/static/3bac0d6b/favicon.ico" width="48" align="right"><div><strong><a href="https://ci.status.im/job/status-go/job/status-go-manual/302/consoleFull">status-go » status-go-manual #302 Console [Jenkins]</a></strong></div></blockquote>
